packages/hierarchies/src/hierarchies/imodel/FilteringHierarchyDefinition.ts

```diff
@@ -86,6 +102,70 @@ export class FilteringHierarchyDefinition implements HierarchyDefinition {
     return pathsIdentifiersPositionsDictionary;
   }
 
+  private createPathsSimilarityLengthCache(): Dictionary<{ lhsIndex: number; rhsIndex: number }, number> {
+    const compare = (lhs: { lhsIndex: number; rhsIndex: number }, rhs: { lhsIndex: number; rhsIndex: number }) => {
+      // istanbul ignore next
+      if (lhs.lhsIndex === rhs.lhsIndex && lhs.rhsIndex === rhs.rhsIndex) {
+        return 0;
+      }
+      // istanbul ignore next
+      if (lhs.rhsIndex === rhs.lhsIndex && lhs.lhsIndex === rhs.rhsIndex) {
+        return 0;
+      }
+      if (lhs.lhsIndex !== rhs.lhsIndex) {
+        // istanbul ignore next
+        return lhs.lhsIndex > rhs.lhsIndex ? 1 : -1;
+      }
+      // istanbul ignore next
+      return lhs.rhsIndex > rhs.rhsIndex ? 1 : -1;
+    };
+    return new Dictionary<{ lhsIndex: number; rhsIndex: number }, number>(compare);
+  }
+
+  private async getPathsSimilarityLength(lhsIndex: number, rhsIndex: number): Promise<number> {
+    let entry = this._pathsSimilarityLengthCache.get({ lhsIndex, rhsIndex });
+    // istanbul ignore if
+    if (entry !== undefined) {
+      return entry;
+    }
+    let similarityLength = 0;
+    const lhsPath = HierarchyFilteringPath.normalize(this._nodeIdentifierPaths[lhsIndex]);
+    const rhsPath = HierarchyFilteringPath.normalize(this._nodeIdentifierPaths[rhsIndex]);
+    // istanbul ignore next
+    const smallerLength = lhsPath.path.length > rhsPath.path.length ? rhsPath.path.length : lhsPath.path.length;
+    for (let i = 0; i < smallerLength; ++i) {
+      const lhsIdentifier = lhsPath.path[i];
+      const rhsIdentifier = rhsPath.path[i];
+      if (HierarchyNodeIdentifier.isInstanceNodeIdentifier(lhsIdentifier) && HierarchyNodeIdentifier.isInstanceNodeIdentifier(rhsIdentifier)) {
+        if (
+          lhsIdentifier.imodelKey !== rhsIdentifier.imodelKey ||
+          lhsIdentifier.id !== rhsIdentifier.id ||
+          (lhsIdentifier.className !== rhsIdentifier.className &&
+            !(await this._imodelAccess.classDerivesFrom(lhsIdentifier.className, rhsIdentifier.className)) &&
+            !(await this._imodelAccess.classDerivesFrom(rhsIdentifier.className, lhsIdentifier.className)))
+        ) {
+          break;
+        }
+        ++similarityLength;
+        continue;
+      }
+
+      // istanbul ignore next
+      if (HierarchyNodeIdentifier.isGenericNodeIdentifier(lhsIdentifier) && HierarchyNodeIdentifier.isGenericNodeIdentifier(rhsIdentifier)) {
+        if (lhsIdentifier.source !== rhsIdentifier.source || lhsIdentifier.id !== rhsIdentifier.id) {
+          break;
+        }
+        ++similarityLength;
+        continue;
+      }
+      // istanbul ignore next
+      break;
+    }
+    entry = similarityLength;
+    this._pathsSimilarityLengthCache.set({ lhsIndex, rhsIndex }, entry);
+    return entry;
```

Comment:
what's the point of assigning this to `entry`?

```suggestion
    this._pathsSimilarityLengthCache.set({ lhsIndex, rhsIndex }, similarityLength);
    return similarityLength;
```